### DISCUSSION TITLE
Summer Objectives

### DISCUSSION THREAD

**Discussion by jardinetsouffleton** [corail-research/SeaPearl.jl#377] [General]
## Areas of improvement

While SeaPearl is almost ready for the XCSP3 competition, there remains many areas of improvement. The way I see it, there are 5 areas where the project can be improved:

### Documentation
If we are to grow this project, we need to be able to on-board new contributors easily. This means that when they come in, they have access to documentation at multiple levels of abstraction: 

1. **Low-level docs to understand the code** On this front, I think the content is good, but needs a bit of polishing and restructuring. Re-organizing the docs so they are easily understandable will be hugely beneficial for SeaPearl
2. **Posts explaining the theory** For this, we have nothing and it is helpful to draw some attention to the project. Some ideas for blog posts include: 1) An overview of methods combining OR/CP and ML, 2) Why we use GNNs and where they have been applied in CP+ML 3) Explain where SeaPearl fits in the CP+ML landscape 4) A roadmap of developments to come. Of course this list could be vastly expanded, and this would help interested users less familiar with this area of research to gain some understanding of what this project is about.
3. **(Fun and Engaging) Tutorials** As of now, SeaPearlZoo is very useful to build models in classic and learning CP, but lacks the interactive features of a jupyter notebook. For the ML+CO summer school, such a tutorial will be built and we can draw some inspiration from it in order to build new tutorials. At a minimum, we should: 1) Make a jupyter notebook version of all SeaPearlZoo examples 2) Document the sets of hyper-parameters to use for learning CP examples 3) Include some exploratory aspects to tutorials (creating variable/value selection heuristics, hyper-parameter tuning, etc.) 4) Make tutorials aimed at learning how to use SeaPearl -something like a series of "micro-courses" 5) Include visualizations for all tutorials

### Representation Learning
This is a broad category that includes RL, GNNs and the graph representation we use for the CP problems. I see the following things:

1. **PPO** We only have working examples with DQN, but using PPO should yield results that are at least comparable to DQN
2. **Generic Graph Representation** The current graph representation used is not a 1-to-1 mapping from problem to graph, which may lead to adverse effects during the learning process. Using a generic graph representation could lead to better performances...
3. **Building a Knowledge Base** It ties into the need for better documentation, but I thing it is fitting to put it here. Right now, the knowledge of SeaPearl is transmitted largely through an obscure oral tradition and while there is a nice mystique to it, we shouldn't rely on tacit knowledge to build a scientific project like this. We need an automated way to store experiments that work as well as those that don't so that we can get better at this. We need to RUN EXPERIMENTS and RECORD THE OUTCOMES

### Constraint Programming
In order to have good comparisons between what we can learn with SeaPearl and SOTA methods, we need, well, SOTA methods... This means we need to implement new functionalities (heuristics, constraints, metaheuristics maybe?). This is an incomplete list, but we should strive to add, at the very least:

1. **Activity-Based Search** Some development was made and still exists, but was never merged. This is an easy first step
2. **Conflict Ordering** https://www.info.ucl.ac.be/~pschaus/assets/publi/cp2015_cos.pdf
3. **This value selection heuristic** https://link.springer.com/chapter/10.1007/978-3-319-18008-3_8

### Double Heuristic
When we have a good base for representation learning and CP, we will finally be able to think about building a double heuristic -combining variable and value selection heuristics in a single agent. This will require some research and we should start by reading some papers and looking at implementations for learned variable selection heuristics as well as cooperative agents to have a better idea of the design we want to try.

### Housekeeping
All the boring stuff that guarantee the project works as intended:

1. **Updating Julia and Dependencies** A lot of the packages we are using are starting to get old and we should update them to 1) gain access to new functionalities and 2) make sure the project does not fall into a deprecation hell hole
4. **Reduce the Dependencies** Some packages included in the  dependencies are not strictly necessary. For example, the plotting tools do not belong to this project. The parser could also be put into a separate project...
5. **Increase Code Coverage** While the 75% coverage we currently have is good, I think we should aim for at least 80%. We still have sections of the codebase where coverage is severely lacking. For example, the RL section only has a ~51% coverage, which is unacceptable for such an important part of the package -or any part of the package for that matter....

**Comment by qcappart** [comment]
Hello,

Many thanks for curating this list. Here are my thoughts:

(1) About the three points about the documentation:  you can always work on that in parallel with the other tasks. :-)

(2) PPO: Yes, it is a great idea to add it as an alternative of DQN. For testing purpose, you should be able to obtain the same results as the one reported in Tom's last paper.

(3) Generic graph representation: actually the research topic of Leo. :-)

(4) Building a Knowledge Base: totally agree - It will require a brainstorm with the whole to decide how to it properly.

(5) Constraint Programming: Yes, I see it as the next short-term steps.

(6) Housekeeping: like the documentation, continue to do that on a regular basis.

(7) Double-heuristic: although very interesting, I see it as a new M.Sc./Ph.D. research project.
